src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2122,44 +2124,71 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     argList      = argList->Rest();
     GenTree* op3 = argList->Current();
 
-    regNumber op1Reg;
-    regNumber op2Reg;
+    regNumber op1NodeReg = op1->GetRegNum();
+    regNumber op2NodeReg = op2->GetRegNum();
+    regNumber op3NodeReg = op3->GetRegNum();
+
+    GenTree* emitOp1 = op1;
+    GenTree* emitOp2 = op2;
+    GenTree* emitOp3 = op3;
 
-    bool       isCommutative   = false;
     const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    if (op1->isContained() || op1->isUsedFromSpillTemp())
     {
-        // 132 form: op1 = (op1 * op3) + [op2]
-
-        ins    = (instruction)(ins - 1);
-        op1Reg = op1->GetRegNum();
-        op2Reg = op3->GetRegNum();
-        op3    = op2;
+        if (targetReg == op2NodeReg)
+        {
+            std::swap(emitOp1, emitOp2);
+            // op2 = ([op1] * op2) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins = _132form;
+            std::swap(emitOp2, emitOp3);
+        }
+        else
+        {
+            // targetReg == op3NodeReg or targetReg == ?
+            // op3 = ([op1] * op2) + op3
+            // 231 form: XMM1 = (XMM2 * [XMM3]) + XMM1
+            ins = _231form;
+            std::swap(emitOp1, emitOp3);
+        }
     }
-    else if (op1->isContained() || op1->isUsedFromSpillTemp())
+    else if (op2->isContained() || op2->isUsedFromSpillTemp())
     {
-        // 231 form: op3 = (op2 * op3) + [op1]
-
-        ins    = (instruction)(ins + 1);
-        op1Reg = op3->GetRegNum();
-        op2Reg = op2->GetRegNum();
-        op3    = op1;
+        if (targetReg == op3NodeReg)
```

Comment:
I think this needs to be `!copiesUpperBits && (targetReg == op3NodeReg)`

Otherwise, `copiesUpperBits` can be true since `op1` is not `Contained` or `UsedFromSpillTemp` and therefore swapping `emitOp1` isn't correct.